cmd/rpc/sock.go

```diff
@@ -0,0 +1,455 @@
+	// check the map to see if the info exists
+	if sub, found := r.subscriptions[rootChainId]; found {
+		// exit with the height of the root-chain-info
+		return sub.Info.Height
+	}
+	return 0
+}
+
+// GetRootChainInfo() retrieves the root chain info from the root chain 'on-demain'
+func (r *RCManager) GetRootChainInfo(rootChainId, chainId uint64) (rootChainInfo *lib.RootChainInfo, err lib.ErrorI) {
+	// lock for thread safety
+	r.l.Lock()
+	defer r.l.Unlock()
+	// if the root chain id is the same as the info
+	sub, found := r.subscriptions[rootChainId]
+	if !found {
+		// exit with 'not subscribed' error
+		return nil, lib.ErrNotSubscribed()
+	}
+	// get the info
+	info, err := sub.RootChainInfo(0, chainId)
+	if err != nil {
+		return nil, err
+	}
+	// update the info
+	sub.Info = *info
+	// exit with the info
+	return
+}
+
+// GetValidatorSet() returns the validator set from the root-chain
+func (r *RCManager) GetValidatorSet(rootChainId, id, rootHeight uint64) (lib.ValidatorSet, lib.ErrorI) {
+	// if the root chain id is the same as the info
+	sub, found := r.subscriptions[rootChainId]
+	if !found {
+		// exit with 'not subscribed' error
+		return lib.ValidatorSet{}, lib.ErrNotSubscribed()
+	}
+	// if rootHeight is the same as the RootChainInfo height
+	if rootHeight == sub.Info.Height || rootHeight == 0 {
+		// exit with a copy the validator set
+		return lib.NewValidatorSet(sub.Info.ValidatorSet)
+	}
+	// if rootHeight is 1 before the RootChainInfo height
+	if rootHeight == sub.Info.Height-1 {
+		// exit with a copy of the previous validator set
+		return lib.NewValidatorSet(sub.Info.LastValidatorSet)
+	}
+	// warn of the remote RPC call to the root chain API
+	r.log.Warnf("Executing remote GetValidatorSet call with requested height=%d for rootChainId=%d", rootHeight, rootChainId)
+	// execute the remote RPC call to the root chain API
+	return sub.ValidatorSet(rootHeight, id)
+}
+
+// GetOrders() returns the order book from the root-chain
+func (r *RCManager) GetOrders(rootChainId, rootHeight, id uint64) (*lib.OrderBook, lib.ErrorI) {
+	// if the root chain id is the same as the info
+	sub, found := r.subscriptions[rootChainId]
+	if !found {
+		// exit with 'not subscribed' error
+		return nil, lib.ErrNotSubscribed()
+	}
+	// if the root chain id and height is the same as the info
+	if sub.Info.Height == rootHeight {
+		// exit with the order books from memory
+		return sub.Info.Orders, nil
+	}
+	// warn of the remote RPC call to the root chain API
+	r.log.Warnf("Executing remote GetOrders call with requested height=%d for rootChainId=%d", rootHeight, rootChainId)
+	// execute the remote call
+	books, err := sub.Orders(rootHeight, id)
+	// if an error occurred during the remote call
+	if err != nil {
+		// exit with error
+		return nil, err
+	}
+	// ensure the order book isn't empty
+	if books == nil || len(books.OrderBooks) == 0 {
+		// exit with error
+		return nil, lib.ErrEmptyOrderBook()
```

Comment:
should an empty order book really be an error? sounds like a normal situation for new or unused/abandoned chains

Reply:
It doesn't have to be -- but that was previously the pattern

<img width="957" alt="image" src="https://github.com/user-attachments/assets/01d7b218-fd14-49a5-b6fb-8863675a5cc6" />